app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/MessageController.kt

```diff
@@ -71,4 +97,132 @@ object MessageController {
+        } else {
+            getMessagesUidsDelta(mailboxUuid, folderId, previousCursor)
+        }
+
+        with(messagesUids) {
+
+            Log.e("API", "Threads - Added   : ${addedShortUids.count()}")
+            Log.e("API", "Threads - Deleted : ${deletedUids.count()}")
+            Log.e("API", "Threads - Updated : ${updatedMessages.count()}")
+
+            addMessages(addedShortUids, folder, mailboxUuid)
+
+            RealmDatabase.mailboxContent().writeBlocking {
+
+                deleteMessages(deletedUids, realm = this)
+                updateMessages(updatedMessages, mailboxUuid, folderId, realm = this)
+
+                if (cursor != null) FolderController.updateFolder(folderId, this) {
+                    it.lastUpdatedAt = Date().toRealmInstant()
+                    it.cursor = cursor
+                }
+            }
+        }
+
+        // TODO: Do we still need this with the new API routes?
+        // val isDraftFolder = FolderController.getFolder(folderId, this)?.role == FolderRole.DRAFT
+        // if (isDraftFolder) DraftController.cleanOrphans(threadsResult.threads, this)
+    }
+
+    private suspend fun addMessages(shortUids: List<String>, folder: Folder, mailboxUuid: String) {
+        if (shortUids.isNotEmpty()) {
+            val uids = getUniquesUidsInReverse(folder, shortUids)
+            val pageSize = ApiRepository.PER_PAGE
+            // val pageSize = 200 // TODO: Magic number
+            var pageStart = 0
+            while (pageStart < uids.count()) {
+                val pageEnd = min(pageStart + pageSize, uids.count())
+                val page = uids.subList(pageStart, pageEnd)
+                ApiRepository.getMessagesByUids(mailboxUuid, folder.id, page).data?.messages?.let { messages ->
+                    FolderController.updateFolder(folder.id) { folder ->
+                        folder.threads += messages.map { it.toThread(mailboxUuid) }.toRealmList()
+                    }
+                }
+                // TODO: Do we want a delay between each call, to not get blocked by the API?
+                delay(500L)
+                pageStart += pageSize
+            }
+        }
+    }
+
+    private fun deleteMessages(uids: List<String>, realm: MutableRealm) {
+        if (uids.isNotEmpty()) {
+            realm.deleteMessages(getMessages(uids, realm))
+            realm.delete(ThreadController.getThreads(uids, realm))
+        }
+    }
+
+    private fun updateMessages(messageFlags: List<MessageFlags>, mailboxUuid: String, folderId: String, realm: MutableRealm) {
+        messageFlags.forEach {
+            val uid = it.shortUid.toLongUid(folderId)
+            updateMessage(uid, realm) { message ->
+                message.answered = it.answered
+                message.isFavorite = it.isFavorite
+                message.forwarded = it.forwarded
+                message.scheduled = it.scheduled
+                message.seen = it.seen
```

Comment:
I would rename the `it` and use an apply on `message`.

Reply:
Done.